README.md

```diff
@@ -0,0 +1,179 @@
+
+`bleanser` is meant to solve this problem in a data agnostic way, so any export can be converted to a unique 'snapshot', and those can be sorted/compared against each other to find redundant data
+
+## How it works
+
+This has `Normaliser`s for different data sources (see [modules](src/bleanser/modules)), and generally follows a pattern like this:
+
+```python
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Iterator
+
+from bleanser.core.processor import BaseNormaliser
+
+class Normaliser(BaseNormaliser):
+
+    @contextmanager
+    def do_cleanup(self, path: Path, *, wdir: Path) -> Iterator[Path]:
+        # temporarily decompress if the data is stored as compressed on disk
+        with self.unpacked(path=path, wdir=wdir) as upath:
+
+            # a temporary file we write 'clean' data to, that can be easily diffed/compared
+            cleaned = self.unique_file_in_tempdir(upath, wdir)
+
+            # some custom code here per-module that writes to 'clean'
+
+        yield cleaned
+
+
+# this script could be run directly, or if its installed in a module like
+# python3 -m bleanser.modules.smscalls --glob ...
+if __name__ == "__main__":
+    Normaliser.main()
+```
+
+This is **always** acting on the data loaded into memory/temporary files, it is not modifying the files itself. Once it determines an input file can be pruned, it will warn you by default, and you can specify `--move` or `--remove` with the CLI (see below) to remove it.
+
+There are particular normalisers for different filetypes, e.g. [`json`](./src/bleanser/modules/json_new.py), [`xml`](./src/bleanser/modules/xml_clean.py), [`sqlite`](./src/bleanser/core/sqlite.py) which might work if your data is especially basic, but typically this requires subclassing one of those and writing some custom code to 'cleanup' the data, so it can be properly compared/diffed.
+
+### do_cleanup
+
+There are two ways you can think about `do_cleanup` (creating a 'cleaned'/snapshot file) -- by specifying an 'upper' or 'lower' bound:
+
+- upper: specify which data you want to drop, dumping everything else to `cleaned`
+- lower: specify which keys/data you want to keep, e.g. only returning a few keys which uniquely identify events in the data
+
+As an example say you had a JSON export:
+
+```json
+[
+  { "id": 5, "images": [{}], "href": "..." },
+  { "id": 6, "images": [{}], "href": "..." },
+  { "id": 7, "images": [{}], "href": "..." }
+]
+```
+
+When comparing this, you could possibly:
+
+1. Just write the `id` to the file. This is somewhat risky as you don't know if the `href` will always remain the same, so you may be losing data
+1. Write the `id` and the `href`, by specifying those two keys you're interested in
+1. Write the `id` and the `href`, by deleting the `images` key (this is different from 2!)
```

Comment:
should be 2 and 3?

Reply:
markdown automatically renders these as 2 and 3, you dont have to write it out manually

Reply:
aah I see, didn't know! A bit confusing when you read as markdown (especially when referring to the other items), but not a huge deal!